preprocessing/data_science_help_functions.py

```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
+
+    :return:
+            | adversarial_validation_result: Adversarial validation score.
+    """
+
+    # Check if it only one dataframe provided
+    if len(dataframe_dict) != 2:
+        # do nothing and return the original data
+        logger.info("Can't apply adversarial_validation because count of dataframes is not equal to 2")
+        return
+
+    # if 2 dataframe than it will be considered as `train` and `test`
+    train = dataframe_dict[list(dataframe_dict.keys())[0]]
+    test = dataframe_dict[list(dataframe_dict.keys())[1]]
+
+    if len(ignore_columns) > 0:
+        columns_to_use = [x for x in list(test.columns) if x not in ignore_columns]
+        train = train[columns_to_use]
+        test = test[columns_to_use]
+
+    # add identifier and combine
+    train['istrain'] = 1
+    test['istrain'] = 0
+
+    # max_dataframe_length
+    for df in [train, test]:
+        if len(df) > max_dataframe_length:
+            df = df.head(max_dataframe_length)
+
+    # add identifier and combine
+    train['istrain'] = 1
+    test['istrain'] = 0
+    df_joined = pd.concat([train, test], axis=0)
+
+    # convert non-numerical columns to integers
+    df_numeric = df_joined.select_dtypes(exclude=['object', 'datetime'])
+    df_obj = df_joined.select_dtypes(include=['object', 'datetime']).copy()
+
+    for c in df_obj:
+        df_obj[c] = pd.factorize(df_obj[c])[0]
+
+    df_joined = pd.concat([df_numeric, df_obj], axis=1)
+
+    # a new target
+    y = df_joined['istrain']
+    df_joined.drop('istrain', axis=1, inplace=True)
+
+    # train classifier
+    adversarial_validation_result, clf = get_adv_validation_score(df_joined, y)
+
+    # Process result:
+    if adversarial_validation_result > threshold:
+        print(
+            f"WARNING! There is significant difference between {list(dataframe_dict.keys())[0]} and {list(dataframe_dict.keys())[0]}\n"
+            f"datasets in terms of feature distribution. Validation score: {adversarial_validation_result}, threshold: {threshold}")
+        print(f"Top features are: {xgb_important_features(clf)}")
+    else:
+        print(
+            f"There is no significant difference between {list(dataframe_dict.keys())[0]} and {list(dataframe_dict.keys())[0]}\n"
+            f"datasets in terms of feature distribution. Validation score: {adversarial_validation_result}, threshold: {threshold}")
+    return adversarial_validation_result
+
+
+def get_adv_validation_score(df_joined: pd.DataFrame,
+                             y: pd.Series) -> Tuple[float, xgb.sklearn.XGBClassifier]:
+    """ Calculate advisarial validation score based on dataframes and XGBClassifier
+
+    :param DataFrame df_joined: Feature dataframe
+    :param Series y: Target series
+
+    :return:
+            | clf: Trained model
+            | mean of KFold validation results (ROC-AUC scores)
+    """
+
+    skf = StratifiedKFold(n_splits=3, shuffle=True, random_state=44)
```

Comment:
Here is a very rigid parameter definition. I don't know is this work for all cases. I would suggest printing out the xgb_params to the user or to inform him/her about the configuration and tell how to change this. I would suggest extracting this to an external json file and pointing to it and loading using json.load(filename)

Reply:
Simplified it a bit

Reply:
Let us discuss this. I would still write it to an external file